pkg/operator/reports.go

```diff
@@ -354,38 +354,68 @@ func getReportPeriod(now time.Time, logger log.FieldLogger, report *metering.Rep
+// else attempt to determine if we can safely delete this object without breaking
+// other Report custom resources that rely on this particular Report.
 func (op *Reporting) handleExpiredReport(logger log.FieldLogger, report *metering.Report, now time.Time) error {
-	// check if the Report is being deleted already
+	// check if the Report deletion timestamp has already been set, exit early.
 	if report.DeletionTimestamp != nil {
-		logger.Warnf("report was already marked for deletion")
+		// TODO: should we be requeuing here?
+		logger.Warnf("the %s Report in the %s namespace was already marked for deletion", report.Name, report.Namespace)
 		return nil
 	}
-	// check if the Report is past its retention time
-	if reportExpired := isReportExpired(op.logger, report, now); !reportExpired {
+
+	// check if the report we're currently processing has reached the
+	// user-provided retention period. If it hasn't, exit early and
+	// let the call site determine how to handle this Report object.
+	if reportExpired := isReportExpired(logger, report, now); !reportExpired {
 		logger.Debugf("the %s Report in the %s namespace has not yet reached the expiration date", report.Name, report.Namespace)
 		return nil
 	}
-	// check if the Report is used by any other Report or ReportQuery, if not delete it
-	if reportIsNotInput := isReportNotUsedAsInput(logger, report, op); reportIsNotInput {
+	// Determine whether this Report object can be safely deleted.
+	// In the case of something like a roll-up Report, where an
+	// aggregate Report uses another Report object as an input,
+	// we need to ensure that we're not deleting any Report that's
+	// listed as a dependency, even if the user-provided retention
+	// period has passed.
+	reportIsNotInput, err := op.isReportNotUsedAsInput(logger, report)
+	if err != nil {
+		logger.Warnf("failed to determine if the %s Report is listed as a dependency: %v", report.Name, err)
+		return nil
+	}
+	// If the Report is not listed as a dependency for another
+	// custom resource, then attempt to make the client-go API
+	// call to delete this resource. Deleting this resource will
+	// result in the informer picking up this deletion event and will run
+	// op.deleteReport once that event has been picked up in the cache.
+	if reportIsNotInput {
 		newDeletionTimestamp := metav1.Now()
 		report.SetDeletionTimestamp(&newDeletionTimestamp)
-		err := op.meteringClient.MeteringV1().Reports(report.Namespace).
-			Delete(context.TODO(), report.Name, metav1.DeleteOptions{})
+
+		err := op.meteringClient.MeteringV1().Reports(report.Namespace).Delete(context.TODO(), report.Name, metav1.DeleteOptions{})
 		if apierrors.IsNotFound(err) {
-			logger.Infof("report: %s, not deleted because it was not found at time delete attempted", report.Name)
+			logger.Infof("failed to delete the %s Report in the %s namespace but it does not exist anymore", report.Name, report.Namespace)
```

Comment:
The rewrite of the log line reads more like normal English, the problem is when there are null values.
Then the old log line reads like this:
`report: , not deleted because it was not found at time delete attempted`
whereas the rewrite reads like this:
`failed to delete the Report in the namespace but it does not exist anymore`
Unless you're a log reader who has made themselves very familiar with the code base, that latter log line makes you question the programmer's sanity for throwing this log line in the first place. The colon and comma are used to give a hint that something is missing.

Reply:
I think that's a decent situation to consider, but we've essentially performed these validation checks at the call site for this method when we attempt to use the client the object from the cache's metadata key. That said, we already configured the logger for this resource to have the report name and namespace as fields, so this kind of logging may be redundant anyways.

Reply:
I think @pweil- mentioned on his overall recent refactor PR, we could use some dev standards documented in a file in the code base. Things like, new functions need a comment;  always try to format log lines like XXX; we use Go style YYY.  We'll get code and log lines we're all more familiar with that way after awhile.